internal/server/v2/observation/facet.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package observation
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"sort"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	pbv2 "github.com/datacommonsorg/mixer/internal/proto/v2"
+	"github.com/datacommonsorg/mixer/internal/server/ranking"
+	"github.com/datacommonsorg/mixer/internal/server/resource"
+	"github.com/datacommonsorg/mixer/internal/server/stat"
+	"github.com/datacommonsorg/mixer/internal/store"
+	"github.com/datacommonsorg/mixer/internal/store/bigtable"
+	"github.com/datacommonsorg/mixer/internal/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+// Gets the obs response given a list of retrieved bt data rows where the data
+// of each row is a list of facets.
+func GetFacetObsResponse(variables []string, facetBtDataList [][]bigtable.BtRow, variableRowPart int) *pbv2.ObservationResponse {
+	// Init response
+	result := &pbv2.ObservationResponse{
+		ByVariable: map[string]*pbv2.VariableObservation{},
+		Facets:     map[string]*pb.Facet{},
+	}
+	for _, variable := range variables {
+		result.ByVariable[variable] = &pbv2.VariableObservation{
+			ByEntity: map[string]*pbv2.EntityObservation{},
+		}
+		result.ByVariable[variable].ByEntity[""] = &pbv2.EntityObservation{}
+	}
+	// Get the list of facets for each sv
+	svToFacetList := map[string][]*pb.Facet{}
+	for _, btData := range facetBtDataList {
+		for _, row := range btData {
+			sv := row.Parts[variableRowPart]
+			if _, ok := svToFacetList[sv]; !ok {
+				svToFacetList[sv] = []*pb.Facet{}
+			}
+			svToFacetList[sv] = append(svToFacetList[sv], row.Data.(*pb.Facets).GetFacets()...)
+		}
+	}
+	// Go through each list of facets, sort and remove duplicates, and add to
+	// result.
+	for sv, facetList := range svToFacetList {
+		entityObservation := &pbv2.EntityObservation{}
+		sort.Sort(ranking.FacetByRank(facetList))
+		seenFacets := map[string]string{}
```

Comment:
very nit: golang set uses empty struct like `map[string]struct{}{}` for seenFacets

Reply:
done